src/Host/Broker/Impl/Startup/CommonStartup.cs

```diff
@@ -142,16 +156,37 @@ internal class CommonStartup {
             return webHost;
         }
 
+        private static void ServiceExit() {
+            using (var serviceController = new ServiceController(Resources.Text_ServiceName)) {
+                serviceController.Stop();
+            }
+        }
+
         internal static void Exit() {
             _cts.Cancel();
 
             Task.Run(async () => {
-                // Give cooperative cancellation 10 seconds to shut the process down gracefully,
-                // but if it didn't work, just terminate it.
-                await Task.Delay(10000);
-                _logger.LogCritical(Resources.Critical_TimeOutShutdown);
-                Environment.Exit((int)BrokerExitCodes.Timeout);
+                if (IsService) {
+                    ServiceExit();
+                } else {
+                    // Give cooperative cancellation 10 seconds to shut the process down gracefully,
+                    // but if it didn't work, just terminate it.
+                    await Task.Delay(10000);
```

Comment:
(sorry, can't see the change in context anymore)

In the original code, it was done in `Task.Run` after triggering the cancellation token. The idea is that this token should be propagated everywhere to serve as an exit trigger for the whole app. In this case, the wait wouldn't complete because the main thread would exit first.

It looks like we have lost a bunch of places that were doing it somewhere along the line though - I only see it ever being used in one place.

Not sure if we still need it at all.